File: lib/aura/llm/parsers/response_parser.rb
# frozen_string_literal: true

require "json"

module Aura
  module LLM
    module Parsers
      class ResponseParser
        def self.parse(output)
          begin
            raw = output.is_a?(String) ? output.to_s : output.to_s
            obj = output.is_a?(String) ? (safe_json_parse(raw) || raw) : output
            if obj.is_a?(Hash)
              if obj["tool"]
                return { type: "tool_call", tool: obj["tool"], args: normalize_args(obj["args"]), summary: obj["summary"] }
              end

              tc = obj["tool_calls"]
              if tc.is_a?(Array) && tc.any?
                call = tc.first || {}
                tool = call["tool"] || call["name"] || (call["function"] && call["function"]["name"])
                args = call["args"] || call["arguments"] || call["input"] || (call["function"] && call["function"]["arguments"]) || {}
                args = normalize_args(args)
                summary = obj["summary"] || call["summary"] || (args.is_a?(Hash) ? args.delete("summary") : nil)
                # If there is a content field in the main object, it's the thought
                thought = obj["content"] || obj["message"] && obj["message"]["content"]
                return { type: "tool_call", tool: tool, args: args || {}, summary: summary, thought: thought }
              end

              nested = obj.dig("choices", 0, "message", "tool_calls")
              if nested.is_a?(Array) && nested.any?
                call = nested.first || {}
                tool = call["tool"] || call["name"] || (call["function"] && call["function"]["name"])
                args = call["args"] || call["arguments"] || call["input"] || (call["function"] && call["function"]["arguments"]) || {}
                args = normalize_args(args)
                summary = obj["summary"] || call["summary"] || (args.is_a?(Hash) ? args.delete("summary") : nil)
                # The content is usually in message.content
                thought = obj.dig("choices", 0, "message", "content")
                return { type: "tool_call", tool: tool, args: args || {}, summary: summary, thought: thought }
              end
            end
            { type: "text", content: output.to_s }
          rescue StandardError
            { type: "text", content: output.to_s }
          end
        end

        def self.safe_json_parse(s)
          begin
            JSON.parse(s)
          rescue StandardError
            blk = extract_json_block(s)
            blk ? JSON.parse(blk) : nil
          end
        end

        def self.extract_json_block(s)
          if s.include?("```")
            a = s.gsub(/^\s*```json\s*/m, "").gsub(/^\s*```\s*/m, "")
            b = a.gsub(/\s*```\s*$/m, "")
            return b.strip if b.strip.start_with?("{")
          end
          start = s.index("{")
          endi = s.rindex("}")
          if start && endi && endi > start
            return s[start..endi]
          end
          nil
        end

        def self.normalize_args(args)
          if args.nil?
            {}
          elsif args.is_a?(String)
            begin
              parsed = JSON.parse(args)
              parsed.is_a?(Hash) ? parsed : { "value" => parsed }
            rescue StandardError
              { "value" => args }
            end
          elsif args.is_a?(Hash)
            out = args.dup
            out.delete("summary")
            out
          else
            args
          end
        end
      end
    end
  end
end

File: src/core/llm/parsers/responseParser.ts
export interface ToolCallResult {
  type: 'tool_call';
  tool: string;
  args: Record<string, unknown>;
  summary?: string | null;
  thought?: string | null;
  finish_reason?: string | null;
}

export interface TextResult {
  type: 'text';
  content: string;
  thought: string;
  finish_reason?: string | null;
}

export type ParseResult = ToolCallResult | TextResult;

export class ResponseParser {
  public static parse(output: unknown): ParseResult {
    if (output === null || output === undefined) {
      return { type: 'text', content: '', thought: '' };
    }

    // 1. If output is string
    if (typeof output === 'string') {
      const trimmed = output.trim();
      const obj = ResponseParser.safeJsonParse(trimmed);
      if (obj && typeof obj === 'object') {
        const parsed = ResponseParser.parseObject(obj);
        if (!parsed.thought && parsed.type === 'tool_call') {
          const preText = ResponseParser.extractPreJsonText(trimmed);
          if (preText) parsed.thought = preText;
        }
        return parsed;
      }
      return { type: 'text', content: output, thought: output };
    }

    // 2. If output is an object
    if (typeof output === 'object') {
      // Try parsing object directly (tool, tool_calls, choices[0].message.tool_calls)
      const parsedObj = ResponseParser.parseObject(output);
      if (parsedObj.type === 'tool_call') {
        return parsedObj;
      }

      // Check if this is an API response / adapter wrapper containing text content
      const record = output as Record<string, unknown>;
      let textContent: string | null = null;

      if (typeof record.content === 'string' && record.content.trim()) {
        textContent = record.content;
      } else if (Array.isArray(record.choices) && record.choices.length > 0) {
        const msg = (record.choices[0] as Record<string, unknown>)?.message as
          | Record<string, unknown>
          | undefined;
        if (typeof msg?.content === 'string' && msg.content.trim()) {
          textContent = msg.content;
        }
      }

      if (textContent) {
        const subParse = ResponseParser.parse(textContent);
        if (subParse.type === 'tool_call') {
          return subParse;
        }
        return {
          type: 'text',
          content: textContent,
          thought: record.thought ? String(record.thought) : textContent,
        };
      }

      const contentVal =
        record.content !== undefined ? String(record.content) : String(output);
      const thoughtVal =
        record.thought !== undefined ? String(record.thought) : contentVal;
      return {
        type: 'text',
        content: contentVal,
        thought: thoughtVal,
      };
    }

    const str = String(output);
    return { type: 'text', content: str, thought: str };
  }

  public static parseObject(obj: unknown): ParseResult {
    if (!obj || typeof obj !== 'object') {
      return {
        type: 'text',
        content: String(obj || ''),
        thought: String(obj || ''),
      };
    }

    const record = obj as Record<string, unknown>;

    // Case A: { tool: "name", args: {...} }
    const toolVal = record.tool;
    if (toolVal) {
      const argsVal = record.args;
      const summaryVal = record.summary;
      let thoughtVal: unknown =
        record.thought ?? record.content ?? record.message ?? null;
      if (
        thoughtVal &&
        typeof thoughtVal === 'object' &&
        (thoughtVal as Record<string, unknown>).content
      ) {
        thoughtVal = (thoughtVal as Record<string, unknown>).content;
      }
      return {
        type: 'tool_call',
        tool: String(toolVal),
        args: ResponseParser.normalizeArgs(argsVal),
        summary: summaryVal ? String(summaryVal) : null,
        thought: thoughtVal ? String(thoughtVal) : null,
      };
    }

    // Case B: { tool_calls: [...] }
    const tc = record.tool_calls;
    if (Array.isArray(tc) && tc.length > 0) {
      const call = (tc[0] || {}) as Record<string, unknown>;
      const tool =
        call.tool ??
        call.name ??
        (call.function as Record<string, unknown>)?.name;
      let args =
        call.args ??
        call.arguments ??
        call.input ??
        (call.function as Record<string, unknown>)?.arguments ??
        {};
      args = ResponseParser.normalizeArgs(args);
      const summary =
        record.summary ??
        call.summary ??
        (typeof args === 'object'
          ? (args as Record<string, unknown>).summary
          : null);
      if (args && typeof args === 'object') {
        delete (args as Record<string, unknown>).summary;
      }
      const thought =
        record.content ??
        (record.message as Record<string, unknown>)?.content ??
        null;
      return {
        type: 'tool_call',
        tool: String(tool || ''),
        args: (args as Record<string, unknown>) || {},
        summary: summary ? String(summary) : null,
        thought: thought ? String(thought) : null,
      };
    }

    // Case C: { choices: [{ message: { tool_calls: [...] } }] }
    const choices = record.choices;
    if (Array.isArray(choices) && choices.length > 0) {
      const nested = (
        (choices[0] as Record<string, unknown>)?.message as Record<
          string,
          unknown
        >
      )?.tool_calls;
      if (Array.isArray(nested) && nested.length > 0) {
        const call = (nested[0] || {}) as Record<string, unknown>;
        const tool =
          call.tool ??
          call.name ??
          (call.function as Record<string, unknown>)?.name;
        let args =
          call.args ??
          call.arguments ??
          call.input ??
          (call.function as Record<string, unknown>)?.arguments ??
          {};
        args = ResponseParser.normalizeArgs(args);
        const summary =
          record.summary ??
          call.summary ??
          (typeof args === 'object'
            ? (args as Record<string, unknown>).summary
            : null);
        if (args && typeof args === 'object') {
          delete (args as Record<string, unknown>).summary;
        }
        const thought =
          (
            (choices[0] as Record<string, unknown>)?.message as Record<
              string,
              unknown
            >
          )?.content ?? null;
        return {
          type: 'tool_call',
          tool: String(tool || ''),
          args: (args as Record<string, unknown>) || {},
          summary: summary ? String(summary) : null,
          thought: thought ? String(thought) : null,
        };
      }
    }

    const contentVal = record.content;
    const thoughtVal = record.thought ?? contentVal;
    if (contentVal !== undefined || thoughtVal !== undefined) {
      return {
        type: 'text',
        content: contentVal !== undefined ? String(contentVal) : '',
        thought: thoughtVal !== undefined ? String(thoughtVal) : '',
      };
    }

    const serialized =
      typeof obj === 'object' && obj !== null
        ? JSON.stringify(obj)
        : String(obj);
    return {
      type: 'text',
      content: serialized,
      thought: serialized,
    };
  }

  public static extractPreJsonText(s: string): string | null {
    const jsonBlockIndex = s.indexOf('```');
    if (jsonBlockIndex > 0) {
      const preText = s.substring(0, jsonBlockIndex).trim();
      if (preText) return preText;
    }
    const braceIndex = s.indexOf('{');
    if (braceIndex > 0) {
      const preText = s.substring(0, braceIndex).trim();
      if (preText) return preText;
    }
    return null;
  }

  public static safeJsonParse(s: string): unknown {
    try {
      return JSON.parse(s);
    } catch (_e) {
      const blk = ResponseParser.extractJsonBlock(s);
      if (blk) {
        try {
          return JSON.parse(blk);
        } catch (_err) {}
      }
      return null;
    }
  }

  public static extractJsonBlock(s: string): string | null {
    const jsonBlockRegex = /```json\s*([\s\S]*?)\s*```/i;
    const jsonMatch = jsonBlockRegex.exec(s);
    if (jsonMatch?.[1]) {
      const candidate = jsonMatch[1].trim();
      if (candidate.startsWith('{')) {
        return candidate;
      }
    }

    const genericBlockRegex = /```\s*([\s\S]*?)\s*```/g;
    genericBlockRegex.lastIndex = 0;
    while (true) {
      const match = genericBlockRegex.exec(s);
      if (match === null) {
        break;
      }
      const candidate = match[1].trim();
      if (candidate.startsWith('{')) {
        return candidate;
      }
    }

    const start = s.indexOf('{');
    const endi = s.lastIndexOf('}');
    if (start !== -1 && endi !== -1 && endi > start) {
      return s.substring(start, endi + 1);
    }
    return null;
  }

  public static normalizeArgs(args: unknown): Record<string, unknown> {
    if (args === null || args === undefined) {
      return {};
    }
    if (typeof args === 'string') {
      try {
        const parsed = JSON.parse(args);
        return parsed && typeof parsed === 'object'
          ? (parsed as Record<string, unknown>)
          : { value: parsed };
      } catch (_e) {
        return { value: args };
      }
    }
    if (typeof args === 'object') {
      const out = { ...(args as Record<string, unknown>) };
      delete out.summary;
      return out;
    }
    return { value: args };
  }
}
